gc.c: Ensure a gap of one byte before the finaliser table.

jepler opened the pull request:
.. or, for !MICROPY_ENABLE_FINALISER, before the first block of the pool.
    
Closes: adafruit/circuitpython#5021
Closes: micropython/micropython#7716

I tested this with a version of https://github.com/micropython/micropython/issues/7716#issuecomment-906872738 and there were no problems detected. ~~I haven't tested it on our real HW yet.~~ it also fixes the original problem on our real hw

jepler committed:
gc.c: Avoid valgrind false positives.

When you wish to use the valgrind memory analysis tool on micropython,
you can arrange to define MICROPY_DEBUG_VALGRIND to enable use of
special valgrind macros. For now, this only fixes `gc_get_ptr`
so that it never emits the diagnostic "Conditional jump or move depends
on uninitialised value(s)".

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler committed:
gc.c: Ensure a gap of one byte before the finaliser table.

.. or, for !MICROPY_ENABLE_FINALISER, before the first block of the pool.

Closes: adafruit/circuitpython#5021
Closes: micropython/micropython#7116
Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler committed:
Merge remote-tracking branch 'origin/master' into pool-fix-micropython

dpgeorge commented on py/gc.c:
```diff
@@ -118,7 +122,7 @@ void gc_init(void *start, void *end) {
     size_t total_byte_len = (byte *)end - (byte *)start;
     #if MICROPY_ENABLE_FINALISER
-    MP_STATE_MEM(gc_alloc_table_byte_len) = total_byte_len * MP_BITS_PER_BYTE / (MP_BITS_PER_BYTE + MP_BITS_PER_BYTE * BLOCKS_PER_ATB / BLOCKS_PER_FTB + MP_BITS_PER_BYTE * BLOCKS_PER_ATB * BYTES_PER_BLOCK);
+    MP_STATE_MEM(gc_alloc_table_byte_len) = (total_byte_len - 1) * MP_BITS_PER_BYTE / (MP_BITS_PER_BYTE + MP_BITS_PER_BYTE * BLOCKS_PER_ATB / BLOCKS_PER_FTB + MP_BITS_PER_BYTE * BLOCKS_PER_ATB * BYTES_PER_BLOCK);
```
I think this needs a comment in the code.  Also maybe the "1" can be a (macro) constant with a name that helps explain why it's needed (and the comment can be written above the macro definition).

dpgeorge commented on py/gc.c:
```diff
@@ -127,7 +131,7 @@ void gc_init(void *start, void *end) {
     #if MICROPY_ENABLE_FINALISER
     size_t gc_finaliser_table_byte_len = (MP_STATE_MEM(gc_alloc_table_byte_len) * BLOCKS_PER_ATB + BLOCKS_PER_FTB - 1) / BLOCKS_PER_FTB;
-    MP_STATE_MEM(gc_finaliser_table_start) = MP_STATE_MEM(gc_alloc_table_start) + MP_STATE_MEM(gc_alloc_table_byte_len);
+    MP_STATE_MEM(gc_finaliser_table_start) = MP_STATE_MEM(gc_alloc_table_start) + MP_STATE_MEM(gc_alloc_table_byte_len) + 1;
```
This "1" can be the same macro constant used for the other "1" above.

dpgeorge reviewed (commented):
Is it at all possible to make a test for this?

jepler committed:
tests: Add extra gc coverage test.

This test must be separate from the "regular" coverage test, as it
destroys the gc heap, and must exit via exit() rather than returning
to python code.

This test fails due to #7716.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler committed:
gc: Name the gap byte and ensure it is zeroed.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler commented:
I've updated this PR:
 * the 1-byte gap is now named
 * it fixes a second bug that occurred when the gap byte was not initialized (adafruit#5348)
 * it unifies two memset calls into one, slight flash savings
 * an assertion and a test were added.  Before this PR, the test would hit the assertion error in the unix coverage build

jepler commented:
CI is encountering difficulties:
```
Run source tools/ci.sh && ci_code_formatting_setup
'Error reading https://launchpad.net/api/devel/~pybricks/+archive/ubuntu/ppa?ws.op=getSigningKeyData (4 tries): Internal Server Error'
Error: Process completed with exit code 1.
```

jepler commented:
.. and the Windows (appveyor) build is failing with almost every test saying "FATAL: uncaught NLR", which could be some new problem I've introduced even though no effect on nlr was intended.

jepler committed:
Merge branch 'test-for-7716' into pool-fix-micropython

dpgeorge commented:
I have rebased this here: https://github.com/dpgeorge/micropython/tree/py-gc-ensure-one-byte-after-atb

dpgeorge commented on py/gc.c:
```diff
@@ -118,7 +128,8 @@ void gc_init(void *start, void *end) {
+    size_t gc_alloc_table_byte_len = (total_byte_len - ALLOC_TABLE_GAP_BYTE) * MP_BITS_PER_BYTE / (MP_BITS_PER_BYTE + MP_BITS_PER_BYTE * BLOCKS_PER_ATB / BLOCKS_PER_FTB + MP_BITS_PER_BYTE * BLOCKS_PER_ATB * BYTES_PER_BLOCK);
+    MP_STATE_MEM(gc_alloc_table_byte_len) = gc_alloc_table_byte_len;
     #else
     MP_STATE_MEM(gc_alloc_table_byte_len) = total_byte_len / (1 + MP_BITS_PER_BYTE / 2 * BYTES_PER_BLOCK);
```
I think this should be `(total_byte_len - ALLOCA_TABLE_GAP_BYTE) ...`

dpgeorge replied:
I fixed this during merge.

dpgeorge commented on ports/unix/coverage.c:
```diff
@@ -650,4 +651,29 @@ STATIC mp_obj_t extra_coverage(void) {
+        heap_size++;
+    }
+    mp_printf(&mp_plat_print, "# end coverage.c gc test\n");
+    exit(0);
```
This is quite brutal, to exit so abruptly!  I wonder if there's a better way to do it...

dpgeorge replied:
I don't think this is a good way to do it because it prevents coverage/testing of the tear down and exiting of the executable.

Two options I can think of:
1. save and then restore the `mp_state_mem_t` struct in `mp_state_ctx_t` at the start/end of this function, which should restore the original heap on exit of this function
2. run this code at the start or end of main() (probably at the start, just before `gc_init()`).

dpgeorge replied:
I did option (1) during merge.

dpgeorge commented:
I rebased this, fixed up things as described above, and merged in 84071590b3d76a647fe9cd9055a089605fedc219 through d75ff422972189232345108ecd2037ac148acda1

dpgeorge commented:
Thanks @jepler , this is an important fix.

jepler commented:
Thanks! And thank you for fixing the test to be more appropriate
